lightning-graph-sync/src/processing.rs

```diff
@@ -0,0 +1,539 @@
+use std::cmp::max;
+use std::io;
+use std::io::Read;
+
+use bitcoin::BlockHash;
+use bitcoin::secp256k1::PublicKey;
+
+use lightning::ln::msgs::{
+	DecodeError, ErrorAction, LightningError, OptionalField, UnsignedChannelUpdate,
+};
+use lightning::routing::network_graph;
+use lightning::util::ser::{BigSize, Readable};
+
+use crate::error::GraphSyncError;
+use crate::GraphSyncError::ProcessingError;
+
+/// Maximum number of distinct node IDs, or public keys, that may be included in an update.
+/// This constraint is necessary to avoid malicious updates being able to trigger excessive
+/// memory allocation.
+const MAX_NODE_ID_COUNT: u32 = 50_000;
+const MAX_ANNOUNCEMENT_COUNT: u32 = 150_000;
+const MAX_UPDATE_COUNT: u32 = 250_000;
+
+/// Update network graph from binary data
+///
+/// `network_graph`: network graph to be updated
+///
+/// `update_data`: `&[u8]` binary stream that comprises the update data
+pub fn update_network_graph(
+	network_graph: &network_graph::NetworkGraph,
+	update_data: &[u8],
+) -> Result<(), GraphSyncError> {
+	let mut read_cursor = io::Cursor::new(update_data);
+	read_network_graph(&network_graph, &mut read_cursor)
+}
+
+pub(crate) fn read_network_graph<R: Read>(
+	network_graph: &network_graph::NetworkGraph,
+	mut read_cursor: &mut R,
+) -> Result<(), GraphSyncError> {
+	let mut prefix = Vec::with_capacity(4);
+	prefix.resize(4, 0);
+	read_cursor.read_exact(&mut prefix)?;
+
+	let _deserialize_unsigned = match &prefix[..] {
+		&[76, 68, 75, 1] => true,
+		_ => {
+			return Err(GraphSyncError::ProcessingError(
+				"Prefix must equal 'LDK' and the version byte 1".to_string(),
+			));
+		}
+	};
+
+	let chain_hash: BlockHash = Readable::read(read_cursor)?;
+	let latest_seen_timestamp: u32 = Readable::read(read_cursor)?; //
+
+	let node_id_count: u32 = Readable::read(read_cursor)?;
```

Comment:
Seems odd we speak generally about the types of top-level fields but don't do the same for fields of inner structs.

Reply:
Are you referring to the readme, Jeff? I wanted to leave in some top-level types to avoid confusion and because they're not subject to change. Block hash will always be 32 bytes. Timestamp will be 32 bytes until 2038, but by then we'll hopefully have eltoo anyway.

Reply:
Yeah, though, I'm referring more to the inconsistency of, say, not giving the type for node id count but giving the types for every field within `CustomChannelAnnouncements` and `CustomChannelUpdate` sections. I'm more inclined to give all the types if the `README` is intended to describe the encoding.

I guess it's more a question if we want the encoding in the `README`. I think having it somewhere is valuable, even if in code-level comments. Just seems odd to remove a few types yet leave all the nitty-gritty details of, say, `CustomChannelUpdate`'s layout.